README
On Debian Lenny, pkgconfig does not work because of a missing file
in the fitsio library. You can solve this by replacing the
MKOCT_FLAGS = $(shell pkg-config --libs cfitsio) with
MKOCT_FLAGS = -lcfitsio
in src/Makefile



NEWS
Version 1.0.7+, released 2019-XX-XX:
===================================

 * Use configure to check for octave version specific function calls

 * add/install BIST tests

 * add low level fits interface

Version 1.0.7, released 2015-06-10:
===================================
 * Allow for extension in read_fits_image( filename, extension ) being zero to read the 
   first image of a file with multiple extension.
   read_fits_image( filename, 0 ) is equivalent to read_fits_image( filename ) and
   read_fits_image( filename[0] ).

 
Version 1.0.6, released 2014-06-10:
===================================
 * Get cflags from pkg-config for compilation. No changes in module sources.


Version 1.0.5, released 2014-04-26:
===================================
 * Don't duplicate data in save_fits_image_multi and save_fits_image_multi_ext


Version 1.0.4, released 2014-03-11:
===================================
 * read_fits_image can now be compiled with Octave 3.0, 3.2, 3.4, 3.6, and 3.8


Version 1.0.3, released 2013-07-03:
===================================
 * read_fits_image fixed crash if NAXIS=1


Version 1.0.2, released 2012-07-24:
===================================
 * read_fits_image can now read images with more than 3 dimensions
   (maximum NAXIS is now 999, was 3)
 * read_fits_image can now be compiled with Octave 3.0, 3.2, 3.4, and 3.6



inst/fitsread.m
## Copyright (C) 2019 Dirk Schmidt <fs@dirk-schmidt.net>
##
## This program is free software; you can redistribute it and/or modify it under
## the terms of the GNU General Public License version 3 as published by the 
## Free Software Foundation.
##
## This program is distributed in the hope that it will be useful, but WITHOUT
## ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
## FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
## details.
##
## You should have received a copy of the GNU General Public License along with
## this program; if not, see <http://www.gnu.org/licenses/>.

## -*- texinfo -*-
## @deftypefn  {Function File} {[@var{image}, @var{header}] =} fitsread (@var{filename},[@var{hdu} [, @var{convert}]])
##
## Read image data from FITS file @var{filename}.
##
## The first two dimensions of the image array get transposed similar to Matlab's fitsread().
## This function is not fully compatible with the parameters in Matlab's fitsread().
##
## @seealso{read_fits_image} for the parameters of this function.
## @end deftypefn

function [img,header] = fitsread( filename, varargin )

  if( nargin < 1 || nargin > 3 )
    print_usage ();
  elseif( ! ischar(filename) )
    error( "fitsread: FILENAME must be a string" )
  endif

  bitpix = [];
  hdu = [];
  for idx = 1:numel(varargin)
    opt = varargin{idx};
    if ( isscalar (opt) )
      hdu = opt;
    elseif( ischar(opt) )
      bitpix = opt;
    else
      error ("fitsread: unrecognized option of class %s", class (opt))
    end
  end

  if( ! isempty(hdu) && isempty(bitpix) )
    [img,header] = read_fits_image( filename, hdu );
  elseif( ! isempty(hdu) && ! isempty(bitpix) )
    [img,header] = read_fits_image( filename, hdu, bitpix );
  elseif( isempty(hdu) && ! isempty(bitpix) )
    [img,header] = read_fits_image( filename, bitpix );
  else
    [img,header] = read_fits_image( filename );
  end

  dims = 1:ndims(img);
  dims(1:2) = dims(2:-1:1);
  img = permute( img, dims );
end


inst/fitswrite.m
## Copyright (C) 2019 Dirk Schmidt <fs@dirk-schmidt.net>
##
## This program is free software; you can redistribute it and/or modify it under
## the terms of the GNU General Public License version 3 as published by the 
## Free Software Foundation.
##
## This program is distributed in the hope that it will be useful, but WITHOUT
## ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
## FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
## details.
##
## You should have received a copy of the GNU General Public License along with
## this program; if not, see <http://www.gnu.org/licenses/>.

## -*- texinfo -*-
## @deftypefn  {Function File} {} fitswrite (@var{image}, @var{filename}, [ @var{bits_per_pixel}, @var{header}] )
##
## Write @var{image} to FITS file @var{filename}.
##
## The first two dimensions of the image array get transposed, similar to Matlab's fitswrite().
## This function is not fully compatible with the parameters in Matlab's fitswrite().
##
## @seealso{save_fits_image}
## @end deftypefn

function fitswrite( img, filename, varargin )

  if (nargin < 2 || nargin > 4)
    print_usage ();
  elseif( ! ( isreal(img) && ! issparse(img) && ! isempty(img) ) )
    error( "fitswrite: IMAGE must be a real n-dimensional array" )
  elseif( ! ischar(filename) )
    error( "fitswrite: FILENAME must be a string" )
  end

  bitpix = [];
  header = [];
  for idx = 1:numel (varargin)
    opt = varargin{idx};
    if (ischar (opt) || isscalar (opt))
      bitpix = opt;
    elseif (iscell (opt) )
      header = opt;
    else
      error ("fitswrite: unrecognized option of class %s", class (opt))
    end
  end


  dims = 1:ndims(img);
  dims(1:2) = dims(2:-1:1);
  writebuf = permute( img, dims );
  save_fits_image( filename, writebuf, bitpix, header );
end



src/save_fits_image.cc
// Copyright (C) 2009-2015 Dirk Schmidt <fs@dirk-schmidt.net>
//
// This program is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation; either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program; if not, see <http://www.gnu.org/licenses/>.


#include <iostream>
#include <sstream>
#include <octave/oct.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

extern "C"
{
#include "fitsio.h"
}

static bool any_bad_argument( const octave_value_list& args );

DEFUN_DLD( save_fits_image, args, nargout,
"-*- texinfo -*-\n\
     @deftypefn {Function File}  save_fits_image(@var{filename}, @var{image}, @var{bit_per_pixel})\n\
     Write @var{IMAGE} to FITS file @var{filename}.\n\n\
     Datacubes will be saved with NAXIS=3.\n\n\
     The optional parameter @var{bit_per_pixel} specifies the data type of the pixel values. Accepted string values are BYTE_IMG, SHORT_IMG, LONG_IMG, LONGLONG_IMG, FLOAT_IMG, and DOUBLE_IMG (default). Alternatively, corresponding numbers may be passed, i.e. 8, 16, 32, 64, -32, and -64.\n\n\
     Use a preceding exclamation mark (!) in the filename to overwrite an existing file.\n\n\
     Lossless file compression can be used by adding the suffix '.gz' to the filename.\n\n\
     @seealso{save_fits_image_multi_ext, read_fits_image}\n\
     @end deftypefn")
{
  if ( any_bad_argument(args) )
    return octave_value_list();

  octave_value fitsimage;
  std::string outfile = args(0).string_value ();


  const NDArray image = args(1).array_value();
  dim_vector dims = image.dims();
  int num_axis = dims.length();
  OCTAVE_LOCAL_BUFFER ( long int, sz_axes, num_axis );
  long int len = 1;
  for( int i=0; i<num_axis; i++ )  
  {
    sz_axes[i] = dims(i);
    len *= dims(i);
  }

  int bitperpixel = DOUBLE_IMG;
  if( 3 == args.length() )
  {
    if( args(2).is_string() )
    {
      if( args(2).string_value() == "BYTE_IMG" )
        bitperpixel = BYTE_IMG;
      else if( args(2).string_value() == "SHORT_IMG" )
        bitperpixel = SHORT_IMG;
      else if( args(2).string_value() == "LONG_IMG" )
        bitperpixel = LONG_IMG;
      else if( args(2).string_value() == "LONGLONG_IMG" )
        bitperpixel = LONGLONG_IMG;
      else if( args(2).string_value() == "FLOAT_IMG" )
        bitperpixel = FLOAT_IMG;
      else if( args(2).string_value() == "DOUBLE_IMG" )
        bitperpixel = DOUBLE_IMG;
      else
      {
        fprintf( stderr, "Invalid string value for 'bit_per_pixel': %s\n", args(2).string_value().c_str() );
        return octave_value_list();
      }
    }
    else if( args(2).is_scalar_type() )
    {
      double val = args(2).double_value();
      if( (OCTAVE__D_NINT( val ) ==  val) )
      {
        if( BYTE_IMG == val )
          bitperpixel = BYTE_IMG;
        else if( SHORT_IMG == val )
          bitperpixel = SHORT_IMG;
        else if( LONG_IMG == val )
          bitperpixel = LONG_IMG;
        else if( LONGLONG_IMG == val )
          bitperpixel = LONGLONG_IMG;
        else if( FLOAT_IMG == val )
          bitperpixel = FLOAT_IMG;
        else if( DOUBLE_IMG == val )
          bitperpixel = DOUBLE_IMG;
        else
        {
          fprintf( stderr, "Invalid numeric value for 'bit_per_pixel': %f\n", val );
          return octave_value_list();
        }
      }
    }
    else
    {
      fprintf( stderr, "Third parameter must be a valid string or a valid scalar value.\nSee 'help save_fits_image' for valid values.\n" );
      return octave_value_list();
    }
  }

  int status=0; // must be initialized with zero (I consider this to be a bug in libcfitsio).
                // status seems not to be set to zero after successful API calls

  // Open FITS file
  fitsfile *fp;
  if ( fits_create_file( &fp, outfile.c_str(), &status) > 0 )
  {
      fprintf( stderr, "Could not open file %s.\n", outfile.c_str() );
      fits_report_error( stderr, status );
      return octave_value_list();  
  }

  long fpixel = 1;
  if( fits_create_img( fp, bitperpixel, num_axis, sz_axes, &status ) > 0 )
  {
    fprintf( stderr, "Could not create HDU.\n" );
    fits_report_error( stderr, status );
    return octave_value_list();
  }

  double * datap = const_cast<double*>( image.fortran_vec() );
  if( fits_write_img( fp, TDOUBLE, fpixel, len, datap , &status ) > 0 )
  {
    fprintf( stderr, "Could not write image data.\n" );
    fits_report_error( stderr, status );
    return octave_value_list();
  }


  // Close FITS file
  if( fits_close_file(fp, &status) > 0 )
  {
      fprintf( stderr, "Could not close file %s.\n", outfile.c_str() );
      fits_report_error( stderr, status );
  }

  return octave_value_list();
}

static bool any_bad_argument( const octave_value_list& args )
{
  if ( args.length() < 2 || args.length() > 3 )
  {
    error( "save_fits_image: number of arguments - expecting save_fits_image( filename, image ) or save_fits_image( filename, image, bitsperpixel )" );
    return true;
  }

  if( !args(0).is_string() )
  {
    error( "save_fits_image: filename (string) expected for first argument" );
    return true;
  }


  return false;
}

#if 0
%!shared testfile
%! testfile = tempname();

%!error <save_fits_image: number of arguments> save_fits_image()

%!error <save_fits_image: number of arguments> save_fits_image(1)

%!error <save_fits_image: filename> save_fits_image(1,1)

%!test
%! data = [ 1, 2, 3; 4,5,6; 1, 7, 11 ];
%! save_fits_image(testfile, data);
%! rd=read_fits_image(testfile);
%! assert(size(rd, 1), 3);
%! assert(size(rd, 2), 3);
%! assert(data, rd)

%! if exist (testfile, 'file')
%!   delete (testfile);
%! endif
#endif



src/fitsinfo.cc
#include <iostream>
#include <sstream>
#include <ctype.h>
#include <octave/oct.h>
#include <octave/version.h>
#include <octave/file-info.h>

extern "C"
{
#include "fitsio.h"
}

static int
get_bin_format (const std::string &coltype, std::string &type, int &len)
{
  // numbers at start
  if(::isdigit (coltype[0]))
    len = atoi (coltype.c_str());
  else if (coltype [0] == 'P')
    len = 2;
  else
    len = 1;

  // skip the numbers to get to data type
  int idx = 0;
  for (idx=0; idx<coltype.length (); idx++)
    {
      if (! ::isdigit(coltype[idx]) && coltype[idx] != 'P') break;
    }

  std::string ftype = coltype.substr (idx);
  if (ftype == "") ftype = "A";

  switch (ftype[0])
    {
      case 'B':
        type = "uint8";
        break;
      case 'D':
        type = "double";
        break;
      case 'J':
        type = "int32";
        break;
      case 'I':
        type = "int16";
        if (coltype[0] == 'P')
          type = "int32";
        break;
      case 'E':
        type = "single";
        break;
      case 'X':
        if (len <= 8)
          type = "bit8";
        else if (len <= 16)
          type = "bit16";
        else if (len <= 32)
          type = "bit32";
        else
          type = "bit64";
        break;
      case 'C':
        type = "single complex";
        break;
      case 'M':
        type = "double complex";
        break;
      default:
        type = "char";
    }

  return ftype[0];
}


static int
get_ascii_format (const std::string &coltype, std::string &type, int &len)
{

  switch(coltype[0])
    {
      case 'I':
      case 'O':
      case 'Z':
        type = "integer";
        break;
      case 'F':
      case 'E':
      case 'G':
        type = "single";
        break;
      case 'D':
        type = "double";
        break;
      default:
        type = "char";
    }

  len = atoi (coltype.substr(1).c_str());

  return coltype[0];
}

DEFUN_DLD( fitsinfo, args, nargout,
"-*- texinfo -*-\n \
@deftypefn {Function File} {[@var{info}]} = fitsinfo(@var{filename})\n \
Read information about fits format file\n \
@end deftypefn")
{
  octave_value_list retval;  // create object to store return values

  if ( args.length() == 0)
    {
      print_usage ();
      return octave_value();
    }
  if ( args.length() != 1 || !args(0).is_string() )
    {
      error( "fitsinfo: filename (string) expected as only argument" );
      return octave_value();
    }

  std::string infile = args(0).string_value ();


  // Open FITS file and position to first HDU containing an image
  int status=0;
  fitsfile *fp;
  if ( fits_open_image( &fp, infile.c_str(), READONLY, &status) > 0 )
    {
      fits_report_error( stderr, status );
      error("Could not open file %s.", infile.c_str());
      return octave_value();
    }

  // map value that will be the info result
  octave_map om;

  // get file info
  octave::file_info stat(infile);

  om.assign("filename", octave_value(infile));
  om.assign("filesize", octave_value(stat.size()));
  om.assign("filemoddate", octave_value(stat.timestamp().ctime()));

  string_vector contents;
  int hdupos, hdutype;
  double offset;

  fits_get_hdu_num(fp, &hdupos);


  for(; !status; hdupos++)
    {
      octave_quit();
  
      octave_map hdu;

      fits_get_hdu_type(fp, &hdutype, &status);

      OFF_T headstart;
      OFF_T datastart;
      OFF_T dataend;
      fits_get_hduoff (fp, &headstart, &datastart, &dataend, &status);

      double datasize = (dataend - datastart);

      // read keywords (common for all types)
      int num_keys, key_pos;
      char card[FLEN_CARD]; 
      char keyname[FLEN_KEYWORD], value[FLEN_VALUE], comment[FLEN_VALUE];

      if( fits_get_hdrpos( fp, &num_keys, &key_pos, &status) > 0 )
        {
          fits_report_error( stderr, status );
          error( "Could not get number of header keywords" ) ;
          return octave_value();
        }

      Cell key_matrix(num_keys, 3);
      int key_count =  0;
      double slope = 1.0;
      double intercept = 0.0;
      double axis1 = 1;
  
      for( int i = 1; i <= num_keys; i++ ) // get the keywords
        { 
          if ( fits_read_record( fp, i, card, &status ) )
            {
              fits_report_error( stderr, status );
              error( "Could not read header keyword" );
              return octave_value ();
            }
          else
            {
              int len;
              char ktype;
              fits_get_keytype(card, &ktype, &status);
              fits_get_keyname(card, keyname, &len, &status);
              fits_parse_value(card, value, comment, &status);
       
              if (status == 0)
                {
                  if(strcmp(keyname, "BZERO") == 0) intercept = atof(value);
                  if(strcmp(keyname, "BSCALE") == 0) slope = atof(value);
                  if(strcmp(keyname, "NAXIS1") == 0) axis1 = atof(value);
                } 
              else if(status == VALUE_UNDEFINED)
                { 
                  status = 0;
                  keyname[0] = '\0';
                  value[0] = '\0';
                  comment[0] = '\0';
                }

              key_matrix(key_count, 0) = octave_value(std::string(keyname));
              key_matrix(key_count, 1) = octave_value(std::string(value));
              key_matrix(key_count, 2) = octave_value(std::string(comment));
              key_count ++;
            }

        }

      hdu.assign("intercept", octave_value(intercept));
      hdu.assign("slope", octave_value(slope));
      hdu.assign("keywords", octave_value(key_matrix));

      if(hdutype == IMAGE_HDU)
        {
          int bits_per_pixel, num_axis;
          int const MAXDIM=999; // max number supported by FITS standard

          std::vector<long> sz_axes(MAXDIM,0);
          if( fits_get_img_param( fp, sz_axes.size(), &bits_per_pixel,
                &num_axis, sz_axes.data(), &status) > 0 )
            {
              fits_report_error( stderr, status );
              error("Could not get image information");
            }

          std::string datatype = "";
          if(bits_per_pixel == BYTE_IMG) datatype = "uint8";
          if(bits_per_pixel == SHORT_IMG) datatype = "uint16";
          if(bits_per_pixel == LONG_IMG) datatype = "uint32";
          if(bits_per_pixel == LONGLONG_IMG) datatype = "uint64";
          if(bits_per_pixel == FLOAT_IMG) datatype = "single";
          if(bits_per_pixel == DOUBLE_IMG) datatype = "double";
          hdu.assign("datatype", octave_value(datatype));

          if(bits_per_pixel < 0)
            datasize = -bits_per_pixel;
          else
            datasize = bits_per_pixel;

          Matrix dv(1, num_axis);
          for ( int i=0; i<num_axis; i++ )
            {
              dv(0, i) = sz_axes[i];
              datasize *= sz_axes[i];
            }
          if (num_axis >= 2)
            {
              // matlab shows axis as Y, X, .... so need swap the first 2
              dv(0, 0) = sz_axes[1];
              dv(0, 1) = sz_axes[0];
            }

          datasize /= 8;

          hdu.assign("size", octave_value(dv));
          Matrix missing(0,0);
          hdu.assign("missingdatavalue", octave_value(missing));

        }
      else
      {
        // ascii or bin table data
        long nrows;
        int ncols;
        fits_get_num_rows(fp, &nrows, &status);
        fits_get_num_cols(fp, &ncols, &status);

        hdu.assign("rows", octave_value(nrows));
        hdu.assign("nfields", octave_value(ncols));


        if(hdutype == ASCII_TBL)
          {
            // ascii table method of getting size
            int linesize = 1;
            Matrix fieldwidth(1, ncols);
            Matrix fieldpos(1, ncols);
            Cell fieldformat(1, ncols);
            Cell fieldprecision(1, ncols);
            Cell fieldintercept(1, ncols);
            Cell fieldslope(1, ncols);
            Cell fieldmissing(1, ncols);
 
            char keyname[FLEN_KEYWORD], colname[FLEN_VALUE];
            char coltype[FLEN_VALUE];

            for (int i = 1; i <= ncols; i++)
              {
                double fc;
                fits_make_keyn ("TBCOL", i, keyname, &status);
                fits_read_key (fp, TDOUBLE, keyname, &fc, NULL, &status);
                if(status == KEY_NO_EXIST)
                  {
                    status = 0;
                    fc = 0;
                  }

                double tmp;
                fits_make_keyn ("TSCAL", i, keyname, &status);
                fits_read_key (fp, TDOUBLE, keyname, &tmp, NULL, &status);
                if (status == KEY_NO_EXIST)
                  {
                    status = 0;
                    tmp = 1;
                  }
                fieldslope(0,i-1) = octave_value(tmp);

                fits_make_keyn ("TZERO", i, keyname, &status);
                fits_read_key (fp, TDOUBLE, keyname, &tmp, NULL, &status);
                if (status == KEY_NO_EXIST)
                  {
                    status = 0;
                    tmp = 0;
                  }
                fieldintercept(0,i-1) = octave_value(tmp);

                fits_make_keyn ("TNULL", i, keyname, &status);
                fits_read_key (fp, TSTRING, keyname, &coltype, NULL, &status);
                if(status == KEY_NO_EXIST)
                  {
                    status = 0;
                    coltype[0] = '\0';
                    fieldmissing(0,i-1) = octave_value(Matrix());
                  }
                else
                  {
                    int j = 0;
                    // we have any chars
                    for(j=0;coltype[j]!='\0' && coltype[j] == ' ';j++) {}

                    if(j == strlen(coltype))
                      coltype[0] = '\0';

                    fieldmissing(0,i-1) = octave_value(coltype);
                  }

                fits_make_keyn("TFORM", i, keyname, &status);
                fits_read_key(fp, TSTRING, keyname, coltype, NULL, &status);

                std::string prec;
                int fw; 
                get_ascii_format(coltype, prec, fw);

                fieldwidth(0,i-1) = double(fw);
                fieldformat(0,i-1) = octave_value(coltype);
                fieldprecision(0,i-1) = octave_value(prec);
                fieldpos(0,i-1) = fc;
                linesize = fc + fw;

              }

            datasize = linesize * nrows;
            hdu.assign("rowsize", octave_value(linesize));
            hdu.assign("fieldwidth", octave_value(fieldwidth));
            hdu.assign("fieldprecision", octave_value(fieldprecision));
            hdu.assign("fieldformat", octave_value(fieldformat));
            hdu.assign("fieldpos", octave_value(fieldpos));
            hdu.assign("intercept", octave_value(fieldintercept));
            hdu.assign("slope", octave_value(fieldslope));
            hdu.assign("missingdatavalue", octave_value(fieldmissing));
          }
        else if(hdutype == BINARY_TBL)
          {
            Matrix fieldsize(1, ncols);
            Cell fieldformat(1, ncols);
            Cell fieldprecision(1, ncols);
            Cell fieldintercept(1, ncols);
            Cell fieldslope(1, ncols);
            Cell fieldmissing(1, ncols);

            char keyname[FLEN_KEYWORD], colname[FLEN_VALUE];
            char coltype[FLEN_VALUE];

            for (int i = 1; i <= ncols; i++)
              {

                double tmp;
                fits_make_keyn("TSCAL", i, keyname, &status);
                fits_read_key(fp, TDOUBLE, keyname, &tmp, NULL, &status);
                if (status == KEY_NO_EXIST)
                  {
                    status = 0;
                    tmp = 1.0;
                  }
                fieldslope(0,i-1) = octave_value(tmp);

                fits_make_keyn("TZERO", i, keyname, &status);
                fits_read_key(fp, TDOUBLE, keyname, &tmp, NULL, &status);
                if (status == KEY_NO_EXIST)
                  {
                    status = 0;
                    tmp = 0;
                  }
                fieldintercept(0,i-1) = octave_value(tmp);

                fits_make_keyn("TNULL", i, keyname, &status);
                fits_read_key(fp, TSTRING, keyname, &coltype, NULL, &status);
                if(status == KEY_NO_EXIST)
                  {
                    status = 0;
                    coltype[0] = '\0';
                    fieldmissing(0,i-1) = octave_value(Matrix());
                  }
                else
                  {
                    int j = 0;
                    // we have any chars
                    for (j=0;coltype[j]!='\0' && coltype[j] == ' ';j++) {}

                    if (j == strlen(coltype))
                      coltype[0] = '\0';

                    fieldmissing(0,i-1) = octave_value(coltype);
                  }

                fits_make_keyn("TFORM", i, keyname, &status);
                fits_read_key(fp, TSTRING, keyname, coltype, NULL, &status);

                std::string prec;
                int fw; 
                get_bin_format(coltype, prec, fw);

                fieldsize(0,i-1) = double(fw);
                fieldformat(0,i-1) = octave_value(coltype);

                fieldprecision(0,i-1) = octave_value(prec);
              }

            datasize = axis1 * nrows;
            hdu.assign("rowsize", octave_value(axis1));
            hdu.assign("fieldsize", octave_value(fieldsize));
            hdu.assign("fieldprecision", octave_value(fieldprecision));
            hdu.assign("fieldformat", octave_value(fieldformat));
            hdu.assign("intercept", octave_value(fieldintercept));
            hdu.assign("slope", octave_value(fieldslope));
            hdu.assign("missingdatavalue", octave_value(fieldmissing));
          }
        }  

      // missingdatavalue
      hdu.assign("datasize", octave_value(datasize));
      hdu.assign("offset", octave_value(double(datastart)));

      if (hdupos == 1)
        {
          om.assign("primarydata", octave_value(hdu));
          contents.append(std::string("primary"));
        }
      else
        {
          if(hdupos > 1 && hdutype == IMAGE_HDU)
            {
              char  value[FLEN_VALUE];
              fits_read_key(fp, TSTRING, "XTENSION", value, NULL, &status);
              if(status == KEY_NO_EXIST)
                status = 0;
              else if (strcmp(value, "IMAGE") != 0)
                hdutype = -1;
            }

          if(hdutype == IMAGE_HDU)
            {
              om.assign("image",  octave_value(hdu));
              contents.append(std::string("image"));
            }
          else if(hdutype == BINARY_TBL)
            {
              om.assign("binarytable",  octave_value(hdu));
              contents.append(std::string("binary table"));
            }
          else if(hdutype == ASCII_TBL)
            {
              om.assign("asciitable",  octave_value(hdu));
              contents.append(std::string("ascii table"));
            }
          else
            {
              om.assign("unknown",  octave_value(hdu));
              contents.append(std::string("unknown"));
            }
        }

      fits_movrel_hdu(fp, 1, NULL, &status);
    }

  if (status == END_OF_FILE) status = 0;

  // Close FITS file
  status = 0;
  if( fits_close_file(fp, &status) > 0 )
    {
      fits_report_error( stderr, status );
      error( "Could not close file %s.", infile.c_str() );
      return octave_value ();
    }

  Cell contents_val (1, contents.numel());
  for(int i=0; i<contents.numel(); i++)
    {
      contents_val(0,i) = octave_value(contents[i]);
    }
  om.assign("contents", octave_value(contents_val));

  retval(0) = om;

  return retval;
}

#if 0
%!shared testfile
%! testfile = urlwrite ( ...
%!   'https://fits.gsfc.nasa.gov/nrao_data/tests/pg93/tst0012.fits', ...
%!   tempname() );

%!fail("fitsinfo")

%!fail("fitsinfo(1)")

%!test
%! s=fitsinfo(testfile);
%! assert(s.filename, testfile);
%! assert(s.contents, {"primary", "binary table", "unknown", ...
%!   "image", "ascii table"});
%!
%! assert(s.primarydata.size, [109 102]);
%! assert(s.primarydata.offset, 2880);
%! assert(s.primarydata.datatype, 'single');
%! assert(s.primarydata.intercept, 0);
%! assert(s.primarydata.slope, 1);
%! assert(s.primarydata.datasize, 44472);
%!
%! assert(s.asciitable.offset, 103680);
%! assert(s.asciitable.rows, 53);
%! assert(s.asciitable.rowsize, 59);
%! assert(s.asciitable.nfields, 8);
%! assert(s.asciitable.datasize, 3127);

%!test
%! if exist (testfile, 'file')
%!   delete (testfile);
%! endif

#endif
